### PR TITLE
feat!: Size effects will now work only on components implementing SizeProvider

### PR DESCRIPTION
**Pull request by st-pasha**
# Description

Similar to all other effects, the `SizeEffect` now targets components implementing `SizeProvider`.

Currently the following classes implement `SizeProvider`:
 - `CircleComponent`,
 - `Viewport`,
 - `NineTileBoxComponent`,
 - `SpriteComponent`,
 - `SpriteGroupComponent`,
 - `SpriteAnimationComponent`,
 - `SpriteAnimationGroupComponent`.

Unlike with other effects, `PositionComponent` does NOT implement `SizeProvider`. This is because the components usually don't pay attention to their `size` when rendering, and thus changing their `size` would not create any visual difference, which users assume is a bug. Now this class of bugs will disappear: `SizeEffect` will simply throw an error saying that the target is not a `SizeProvider`.

Breaking change:
 - If users have their own components that they target with `SizeEffect`, they would need to add `implements SizeProvider` to the declaration of their component.

## Checklist

- [x] The title of my PR starts with a [Conventional Commit] prefix (`fix:`, `feat:`, `docs:` etc).
- [x] I have read the [Contributor Guide] and followed the process outlined for submitting PRs.
- [x] I have updated/added tests for ALL new/updated/fixed functionality.
- [x] I have updated/added relevant documentation in `docs` and added dartdoc comments with `///`.
- [x] I have updated/added relevant examples in `examples`.

## Breaking Change

- [x] Yes, this is a breaking change.
- [-] No, this is *not* a breaking change.

## Related Issues

[issue database]: https://github.com/flame-engine/flame/issues
[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Flame Style Guide]: https://github.com/flame-engine/flame/blob/main/STYLEGUIDE.md
[Conventional Commit]: https://conventionalcommits.org

### CODE REVIEW

**Comment by spydon** [comment]
Many common PositionComponents does take size into consideration when rendering as far as I'm aware, so this will be a breaking change for all those ones. Either all the ones that SizeEffect currently works on need to have SizeProvider, or we need to make all PositionComponents take size into consideration.

**Comment by st-pasha** [comment]
Yep. If a component allows its size to be dynamically updated, it marks itself with `implements SizeProvider`. Otherwise, `SizeEffect` will complain about being attached to a component that's not resizable.

File: examples/lib/stories/effects/size_effect_example.dart [hunk 1]
```diff
@@ -15,35 +15,29 @@ class SizeEffectExample extends FlameGame with TapDetector {
     down, depending on its current state.
   ''';
 
-  late RectangleComponent square;
+  late Component shape;
   bool grow = true;
 
   @override
   Future<void> onLoad() async {
-    square = RectangleComponent.square(
```
Comment (spydon) [inline examples/lib/stories/effects/size_effect_example.dart:23]: I wondered how this one could have worked previously, the answer was simple; it didn't. :sweat_smile:

Comment (spydon) [inline examples/lib/stories/effects/size_effect_example.dart:23]: I wondered how this one could have worked previously, the answer was simple; it didn't. :sweat_smile:

**Review by spydon — APPROVED** [review]

**Review by wolfenrain — APPROVED** [review]

**Pull request merged**